### PR TITLE
MySQL: support remaining constraint parts of CREATE/ALTER TABLE

### PR DESCRIPTION
**Pull request by yoichi**
### Brief summary of the change made

Continuing from https://github.com/sqlfluff/sqlfluff/pull/3826, I implemented all the parts related to constraints in CREATE TABLE and ALTER TABLE, including CHECK constraint support.

https://dev.mysql.com/doc/refman/8.0/en/create-table.html
https://dev.mysql.com/doc/refman/8.0/en/alter-table.html

### Are there any other side effects of this change that we should be aware of?

* It reuses TableConstraintSegment in ALTER TABLE ADD CONSTRAINT, the parse structure of ALTER changes.
* It reuses IndexOptionsSegment in ALTER TABLE, where (SECONDARY_)ENGINE_ATTRIBUTE is not allowed. I think it is not the linter's job to check for all syntax errors, and I did to avoid maintaining each for CREATE TABLE and ALTER TABLE.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 1]
```diff
@@ -1050,25 +1060,10 @@ class AlterTableStatementSegment(BaseSegment):
...
                 Sequence(
                     "ADD",
-                    Ref.keyword("UNIQUE", optional=True),
-                    OneOf("INDEX", "KEY", optional=True),
-                    Ref("IndexReferenceSegment"),
-                    Ref("IndexTypeGrammar", optional=True),
-                    Ref("BracketedColumnReferenceListGrammar"),
-                    AnySetOf(
-                        Sequence(
-                            "KEY_BLOCK_SIZE",
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:1063]: Why was this deleted? I don't see similar code in `TableConstraintSegment`.

Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_mysql.py:1063]: Why was this deleted? I don't see similar code in TableConstraintSegment.

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_mysql.py:1063]: With these changes, is it still possible to use `KEY_BLOCK_SIZE`? (Strangely, there is no `.sql` file using this.) Could you please add a test case for it to your PR?

You could use one of the examples from this page: https://dev.mysql.com/doc/refman/5.7/en/innodb-compression-usage.html

Comment (yoichi) [reply src/sqlfluff/dialects/dialect_mysql.py:1063]: > With these changes, is it still possible to use KEY_BLOCK_SIZE? (Strangely, there is no .sql file using this.)

Yes.

> Could you please add a test case for it to your PR?

Added in 251294722ed48b2979577b2250870b1499051e9f
Thanks

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_mysql.py:1088]: Thanks!

**Review by barrywhart — APPROVED** [review]
